include/envoy/buffer/buffer.h

```diff
@@ -18,9 +18,21 @@ namespace Buffer {
 
 /**
  * A raw memory data slice including location and length.
+ * @note The mem_ field is purposely immutable, in support of Buffer::Instance's
+ *       reserve()/commit() idiom:
+ *
+ *          RawSlice slices[num_slices];
+ *          num_slices_used = buffer.reserve(length, slices, num_slices);
+ *          // fill the slices using readv(2)
+ *          // then update the len_ of each slice to indicate the number of bytes filled
+ *          buffer.commit(slices, num_slices_used);
+ *
+ *      Because the mem_ field of each slice is immutable, a buffer's commit() implementation
+ *      can use it as a key to map the slices back to whatever internal data structures
+ *      the buffer implementation is using.
  */
 struct RawSlice {
-  void* mem_ = nullptr;
+  void* const mem_ = nullptr;
```

Comment:
> Actually I just looked over the Buffer interface and it's mostly char*.

What Buffer interface are you referencing? Buffer::Instance and Buffer::OwnedImpl don't appear to use char* in their APIs at all.

Switching from void* to char* in the public-facing interface of the Buffer classes would require a lot of `reinterpret_cast`s. The reason is that most of the protocols Envoy supports are binary, and they need to parse their messages as uint8_t rather than char because the unspecified (implementation-dependent) signedness of char makes it tricky to use safely.

Switching from void* to uint8_t* would still require `reinterpret_cast`s, but I think the amount of affected code would be smaller: just the filters and codecs that currently treat buffers as sequences of characters. Those tend to be legacy protocol implementations, like HTTP/1.

Reply:
Sorry about that; I missed the point that RawSlice is an exposed interface of a concrete struct, used at many call-sites in the codebase. My high level bit is to reduce const_cast in the impl because it requires scrutiny.